src/ZkFold/Symbolic/Data/UInt.hs

```diff
@@ -127,522 +102,417 @@ cast n =
+instance (Symbolic c, KnownNat n, KnownRegisterSize r) => MultiplicativeMonoid (UInt n r c) where
     one = fromConstant (1 :: Natural)
 
-instance (Finite (Zp p), KnownNat n, KnownRegisterSize r) => Semiring (UInt n r (Interpreter (Zp p)))
+instance (Symbolic c, KnownNat n, KnownRegisterSize r) => Semiring (UInt n r c)
 
-instance (Finite (Zp p), KnownNat n, KnownRegisterSize r) => Ring (UInt n r (Interpreter (Zp p)))
-
-instance (Finite (Zp p), KnownNat n, KnownRegisterSize r) => Arbitrary (UInt n r (Interpreter (Zp p))) where
+instance (Symbolic c, KnownNat n, KnownRegisterSize r) => Arbitrary (UInt n r c) where
     arbitrary = do
-        lo <- replicateA (numberOfRegisters @(Zp p) @n @r -! 1) (toss $ registerSize @(Zp p) @n @r)
-        hi <- toss (highRegisterSize @(Zp p) @n @r)
-        return $ UInt $ Interpreter $ V.unsafeToVector (lo <> [hi])
+        lo <- replicateA (numberOfRegisters @(BaseField c) @n @r -! 1) (toss $ registerSize @(BaseField c) @n @r)
+        hi <- toss (highRegisterSize @(BaseField c) @n @r)
+        return $ UInt $ embed $ V.unsafeToVector (lo <> [hi])
         where toss b = fromConstant <$> chooseInteger (0, 2 ^ b - 1)
 
-instance (Finite (Zp p), KnownNat n, KnownRegisterSize r) => Iso (ByteString n (Interpreter (Zp p))) (UInt n r (Interpreter (Zp p))) where
-    from = fromConstant @Natural . toConstant
+instance (Symbolic c, KnownNat n, KnownRegisterSize r, ToConstant (ByteString n c) Natural, FromConstant (c Par1) Natural) => Iso (ByteString n c) (UInt n r c) where
+    from b = UInt $ embed $ V.unsafeToVector [fromConstant @Natural $ toConstant b]
 
-instance (Finite (Zp p), KnownNat n, KnownRegisterSize r) => Iso (UInt n r (Interpreter (Zp p))) (ByteString n (Interpreter (Zp p))) where
-    from = fromConstant @Natural . toConstant
+instance (Symbolic c, KnownNat n, KnownRegisterSize r, FromConstant (c Par1) Natural, ToConstant (ByteString n c) Natural) => Iso (UInt n r c) (ByteString n c) where
+    from b = ByteString $ embed $ V.unsafeToVector [fromConstant @Natural $ toConstant b]
 
---------------------------------------------------------------------------------
+-- --------------------------------------------------------------------------------
 
 instance
-    ( Arithmetic a
+    ( Symbolic c
     , KnownNat n
     , KnownNat k
     , KnownRegisterSize r
     , n <= k
-    , from ~ NumberOfRegisters a n r
-    , to ~ NumberOfRegisters a k r
-    ) => Extend (UInt n r (ArithmeticCircuit a)) (UInt k r (ArithmeticCircuit a)) where
-    extend (UInt ac) = UInt (circuitF solve)
+    , Finite (BaseField с)
+    , ToConstant (Vector (NumberOfRegisters (BaseField c) n r) (BaseField c)) Natural
+    ,FromConstant Natural (Vector (NumberOfRegisters (BaseField c) k r) (BaseField c))
+    -- , BaseField c ~ BaseField c
+    ) => Extend (UInt n r c) (UInt k r c) where
+    extend (UInt x) =  UInt $ symbolicF x (\l -> fromConstant (toConstant l :: Natural)) solve
```

Comment:
Doing bit expansion to extend the number seems not optimal to me. Is there a better solution?

Reply:
In case of `r ~ Auto` -- no, unfortunately, as changing the `n` might change the register size.
However in case of `r ~ Fixed rs`, we can pad with the necessary amount of zeros, for sure.

Reply:
Ok, agree!